packages/crud/src/vaadin-crud-grid.js

```diff
@@ -128,26 +146,59 @@ class CrudGrid extends IncludedMixin(Grid) {
       .replace(/^./, (match) => match.toUpperCase());
   }
 
+  /** @private */
+  __getGroupDepth(column) {
+    if (column === this) {
+      return 0;
+    } else {
+      return 1 + this.__getGroupDepth(column.parentElement);
+    }
+  }
+
   /** @private */
   __createColumn(parent, path) {
-    const col = document.createElement('vaadin-grid-column');
+    const parentGroupDepth = this.__getGroupDepth(parent);
+    // Deduct 1 from the item property depth to account for the value
+    if (parentGroupDepth < this.__itemPropertyDepth - 1) {
+      // Make sure the column has enough groups to match the item property depth
+      const newParent = document.createElement('vaadin-grid-column-group');
+      parent.appendChild(newParent);
+      this.__createColumn(newParent, path);
+      return;
+    }
 
-    col.renderer = (root, column, model) => {
-      root.textContent = path ? this.get(path, model.item) : model.item;
-    };
+    let col;
+    if (!this.noFilter && !this.noSort && !parent.__sortColumnGroup) {
+      // This crud-grid has both a sorter and a filter, but neither has yet been
+      // created => col should become the sorter group column
+      col = document.createElement('vaadin-grid-column-group');
+      col.__sortColumnGroup = true;
+      parent.appendChild(col);
+      // Create the filter column under this sorter group column
+      this.__createColumn(col, path);
+    } else {
+      // In all other cases, col should be a regular column with a renderer
+      col = document.createElement('vaadin-grid-column');
+      parent.appendChild(col);
+      col.renderer = (root, _column, model) => {
+        root.textContent = path ? this.get(path, model.item) : model.item;
+      };
+    }
 
     if (!this.noHead && path) {
+      // Create a header renderer for the column (or column group)
       col.headerRenderer = (root) => {
         const label = this._generateHeader(path);
 
-        if (!this.noSort) {
+        if (col.__sortColumnGroup || (this.noFilter && !this.noSort)) {
+          // The column is either the sorter group column or the root level
+          // sort column (in case a filter isn't used at all) => add the sort indicator
           const sorter = window.document.createElement('vaadin-grid-sorter');
```

Comment:
nit: not related to the PR but let's change this for consistency.

```suggestion
          const sorter = document.createElement('vaadin-grid-sorter');
```